torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -2731,6 +2737,61 @@ class TestPickler:
             torch.distributed.rpc.api._default_pickler is _internal_rpc_pickler
         )
 
+    @dist_init
+    def test_wait_all(self):
+        with rpc.wait_all():
+            self.assertTrue(_thread_local_var.future_list == [])
+            dst = worker_name((self.rank + 1) % self.world_size)
+            fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
+            was_waited_ptr = WasFutureWaitedHelper()
+            print("original was waited ptr type is {0}".format(was_waited_ptr))
+            rpc.rpc_async(dst, set_arg_to_waited, args=(was_waited_ptr,))
```

Comment:
@mrshenli Logically set_arg_to_waited should be updating was_waited_ptr before the corresponding assert statement but this is not happening; its as if set_arg_to_waited is receiving a copy of was_waited_ptr and updating that instead.  

I wanted to ask first do you have a preferred way in mind of testing whether the context manager successfully waited the future?  If not do you off the top of your head know why my expected behavior is not occurring?

If not no worries I'll find the solution, I just don't want to waste more time debugging this if you had something else in mind

Reply:
> I wanted to ask first do you have a preferred way in mind of testing whether the context manager successfully waited the future?

Good point, let me add a `done()` API to future.

> If not do you off the top of your head know why my expected behavior is not occurring?

Because they are operating on different objects. When you pass `was_waited_ptr` to `rpc_async`, the `was_waited_ptr` will be pickled on the caller, and then unpickled on the callee, and then the callee will set it's `was_future_waited` field. So the `was_waited_ptr` would keep intact.